src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -757,7 +758,7 @@ internal static void SetMode(ConsoleHandle consoleHandle, ConsoleModes mode)
             bool result =
                 NativeMethods.ReadConsole(
                     consoleHandle.DangerousGetHandle(),
-                    buffer,
+                    out buffer,
```

Comment:
Not sure why this was changed to an `out` parameter, since a valid object reference is passed in?

Reply:
It come from the method signature https://docs.microsoft.com/en-us/windows/console/readconsole
My intentions were to fix whole signatures, but it’s too much work and I stop this.
I can revert the change too.

Reply:
By pure coincidence, I noticed this change, and I don't think it's correct. See the last comment [here](https://social.msdn.microsoft.com/Forums/en-US/f9047a3d-37a7-45b8-92e6-1355b4f0a7a4/how-does-dllimport-pinvoke-know-how-to-work-with-stringbuilder). The original MSDN mag article seems to be gone, but the relevant part is quoted:

**StringBuilder and Marshaling**
The CLR marshaler has built-in knowledge of the `StringBuilder` type and treats it differently from other types. By default, `StringBuilder` is passed as `[InAttribute, OutAttribute]`. `StringBuilder` is special because it has a `Capacity` property that can determine the size of the required buffer at run time, and it can be changed dynamically. Therefore, during the marshaling process, the CLR can pin `StringBuilder`, directly pass the address of internal buffer used in `StringBuilder`, and allow the contents of this buffer to be changed by native code in place.

To take full advantage of `StringBuilder`, you'll need to follow all of these rules:

1. Don't pass `StringBuilder` by reference (using `out` or `ref`). Otherwise, the CLR will expect the signature of this argument to be `wchar_t **` instead of `wchar_t *`, and it won't be able to pin `StringBuilder`'s internal buffer. Performance will be significantly degraded.
1. Use `StringBuilder` when the unmanaged code is using Unicode. Otherwise, the CLR will have to make a copy of the string and convert it between Unicode and ANSI, thus degrading performance. Usually you should marshal `StringBuilder` as `LPARRAY` of Unicode characters or as `LPWSTR`.
1. Always specify the capacity of `StringBuilder` in advance and make sure the capacity is big enough to hold the buffer. The best practice on the unmanaged code side is to accept the size of the string buffer as an argument to avoid buffer overruns. In COM, you can also use `size_is` in IDL to specify the size.

---
In reply to: [255789200](https://github.com/PowerShell/PowerShell/pull/8847#discussion_r255789200) [](ancestors = 255789200)

Reply:
@jazzdelightsme Thanks! Will revert.